### ISSUE TITLE
[Feature]: Kube Play under Systemd - Default restart policy should be never

### ISSUE THREAD

**Issue by ygalblum** [containers/podman#17036]
### Feature request description

One of the reasons for opening the issue for supporting Kubernetes Jobs: https://github.com/containers/podman/issues/17011, was a discussion about how the responsibilities to restart pods should be split between Podman and systemd when running `kube play` inside a systemd service. The idea was that since Jobs mandate that the `restartPolicy` be set to either `Never` or `OnFailure`, using them with parallelism and completions set to 1, podman will not restart the pod leaving the decision to systemd.

But, maybe a simpler solution would be to use a default value of `Never` for the `restartPolicy` when running under systemd (using the --service-cotainer flag).

**Comment by ygalblum** [comment]
@rhatdan @vrothberg @alexlarsson WDYT?

**Comment by vrothberg** [comment]
After the last conversation, I felt that Kubernetes Jobs would do that.  Changing the restart policy is potentially a breaking change for some users.

**Comment by ygalblum** [comment]
Jobs will do that but mostly because they force the restartPolicy to not be Always.
As for breaking behavior, 4.4 already includes behavior changes when using --service-container, e.g. default log-driver

**Comment by vrothberg** [comment]
> As for breaking behavior, 4.4 already includes behavior changes when using --service-container, e.g. default log-driver

I consider that to be an improvement.  But changing the restart policy can be a big break for users relying on it.

**Comment by rhatdan** [comment]
I thought the agreement was to differentiate between a JOB(Restart=never) and a Deployment (Restart=Always).

**Comment by ygalblum** [comment]
@rhatdan you are right. This is what we discussed. But, then I got to thinking about the reasoning behind this decision and what we actually wanted to achieve.

The issue we discussed was that when running with Restart=Always, Podman will handle the rescheduling while we would like to leave the decision to systemd. Now, while Job disallows Restart=Always and as a result addresses this requirement, there are additional aspects required to support it such as handling of allowed number of failures.

That's why I circled back to what we actually want - when running under systemd, podman should not restart containers and instead let systemd make the decision. For that, the containers' restart policy should be never. The simple solution for this requirement is using Never as the default restart policy when running under systemd. But, as @vrothberg wrote, it is a behavior change that might break some users.

**Comment by vrothberg** [comment]
I am also torn between keeping the status quo and changing the restart policy.  One (major) issue with the current systemd integration is that there is no way (yet?) to propagate the exit status to the service container.  Hence, even if all containers fail miserably and exit non-zero while the service container exits cleanly.

At the time of writing, I feel more comfortable leaving things as they are now but remain open to (new) use cases.  One thing to continue thinking about is definitely why/if/when we want to propagate exit codes to the service container.

**Comment by ygalblum** [comment]
Now I'm a bit confused. IIUC systemd will not get the containers' exit code in any case. So, using Jobs won't help us either, will they?

**Comment by vrothberg** [comment]
> Now I'm a bit confused. IIUC systemd will not get the containers' exit code in any case. So, using Jobs won't help us either, will they?

Not yet.  We discussed the exit-code propagation last week.  Job is the right way to have a one-shot-like experience but the exit-code propagation is still an open question.  We looked at how Kubernetes handles things and IIRC all containers must succeed and if one doesn't there is some exponential back-off with a counter limit.

**Comment by rhatdan** [comment]
Well if there is only one container, then we should propagate.
More then one container, and all fail, propagate.
One or more containers and all succeed, then propagate success.
The only question is what do to with at least one fails and at least one succeeds.
In OneShot, I believe this should be handled as success, We could spice it up by allowing an attribute that told podman to ignore failures for a particular container.

**Comment by ygalblum** [comment]
For sure the question is about the last case. But, I'm not sure I agree with you. In K8S, the pod state is deemed as Error if any of its containers returned an error. As a result, a Job run will be considered as Failed if one of the containers failed and once all retries are done, the job will have a status of failed.
I think this makes more sense since the orchestrator (be it K8S or podman) cannot differentiate between the containers.

**Comment by vrothberg** [comment]
> The only question is what do to with at least one fails and at least one succeeds.
In OneShot, I believe this should be handled as success, We could spice it up by allowing an attribute that told podman to ignore failures for a particular container.

I think we should do what K8s does.

> In K8S, the pod state is deemed as Error if any of its containers returned an error. As a result, a Job run will be considered as Failed if one of the containers failed and once all retries are done, the job will have a status of failed.
I think this makes more sense since the orchestrator (be it K8S or podman) cannot differentiate between the containers.

Exactly that's what we've been discussing/looking at last week :+1:

**Comment by vrothberg** [comment]
> Well if there is only one container, then we should propagate.

I am not sure we should unconditionally.  The propagation should be determined by the "Kind".

Let's first get propagation nailed down for Jobs, then we can consider opting-in for other kinds.
